### PR TITLE
Don't let the UTF8OutPipeReader be static

### LINKED ISSUE #1997: Shells no longer exit, new tabs no longer open (after the first one)
Whoops, I just realized why this was happening. I opened a new tab and the output from a different shell showed up in it.

### PR DESCRIPTION
**Pull request by DHowett-MSFT**
Whoops. This fixes #1997.

### CODE REVIEW

**Comment by german-one** [comment]
Thanks for keeping me updated!

How embarrassing for me. I discussed with @miniksa about a static class instance in `WriteConsoleA` which seems to made me blind for the different preconditions here. However, there is no excuse.

Would you mind to update this line too?
https://github.com/microsoft/terminal/blob/8d52ba0990f89de552b572381adb26b77cee847b/src/types/ut_types/UTF8OutPipeReaderTests.cpp#L124
`UTF8OutPipeReader reader{ outPipe }; // declare an instance of UTF8OutPipeReader`  
Not that this would make any difference here. Just to reconcile it with the productive code.

**Comment by DHowett-MSFT** [comment]
Sure! Thanks for the catch! :smile:

File: src/types/UTF8OutPipeReader.cpp [hunk 1]
```diff
@@ -32,13 +32,13 @@ UTF8OutPipeReader::UTF8OutPipeReader(wil::unique_hfile& outPipe) :
...
     _dwPartialsLen = 0;
 
-    if (dwRead == 0) // quit if no data has been read and no cached data was left over
+    if (!fSuccess) // reading failed (we must check this first, because dwRead will also be 0.)
     {
-        return S_OK;
+        return E_FAIL;
     }
-    else if (!fSuccess) // reading failed
+    else if (dwRead == 0) // quit if no data has been read and no cached data was left over
```
Comment (german-one) [inline src/types/UTF8OutPipeReader.cpp]: Once you update this, consider to make `else if` just an `if`. The `else` is completely meaningless here.

**Comment by german-one** [comment]
Hold on a second
image
Seems updating the Unit Test was a real good idea. I already have a rough idea why this fails. It's likely that we can't work with `std::string_view&` any longer in `UTF8OutPipeReader::Read`.

**Comment by DHowett-MSFT** [comment]
Unless the individual calls to `RunTest` were intended to rely on static state stored in the UTF8OPutPipeReader (partials from one test flowing into the next test), making it non-static was correct.

**Comment by german-one** [comment]
Yes this was definitely correct. I feel responsible for that. Just give me some minutes figuring out what's going wrong ...

**Comment by german-one** [comment]
Having `if (!fSuccess)` before `if (dwRead == 0)` makes it fail. Even though I agree that this order is correct...

**Comment by german-one** [comment]
@DHowett-MSFT Would that be reasonable?
```
    if (!fSuccess) // reading failed
    {
        if (GetLastError() == ERROR_BROKEN_PIPE) // if the pipe has been ended
        {
            return S_OK;
        }

        return E_FAIL;
    }

    if (dwRead == 0) // quit if no data has been read and no cached data was left over
    {
        return S_OK;
    }
```

And would that be redundant in this case?
https://github.com/microsoft/terminal/blob/8d52ba0990f89de552b572381adb26b77cee847b/src/cascadia/TerminalConnection/ConhostConnection.cpp#L202

**Comment by german-one** [comment]
The reason why I originally checked `if (dwRead == 0)` first was because of that piece of the comment:

> // quit if no data has been read **and no cached data was left over**

Now that we escape if `RadFile` failed, we may still have data in the buffer that has been cached before.

**Comment by DHowett-MSFT** [comment]
If there's data in the buffer that was cached, but ReadFile failed, that data can't be used. It's still a partial UTF-8 sequence!

**Comment by german-one** [comment]
Absolutely. So we could just discard it.

**Comment by DHowett-MSFT** [comment]
Okay, this is now a little different. I've cleaned up some of the code and added another _successful_ exit case (reading from a closed pipe is a detectable 0-byte "successful" read.)

**Comment by DHowett-MSFT** [comment]
I switched away from using `unique_hfile` because that's only supposed to be used when a class _owns_ a handle. Holding a reference to a unique_hfile is the same as holding a raw HANDLE!

**Comment by german-one** [comment]
I like this. Looks a little like my proposal above.

File: src/types/UTF8OutPipeReader.cpp [hunk 2]
```diff
@@ -6,19 +6,32 @@
...
+// Arguments:
+//   - strView: on return, populated with successfully-read codepoints.
+// Return Value:
+//   An HRESULT indicating whether the read was successful. For the purposes of this
+//   method, a closed pipe is considered a successful (but false!) read. All other errors
+//   are translated into an appropriate status code.
+//   S_OK for a successful read
+//   S_FALSE for a read on a closed pipe
+//   E_* (anything) for a failed read
 [[nodiscard]] HRESULT UTF8OutPipeReader::Read(_Out_ std::string_view& strView)
```
Comment (german-one) [inline src/types/UTF8OutPipeReader.cpp]: Thanks for adding this!

File: src/types/UTF8OutPipeReader.cpp [hunk 3]
```diff
@@ -27,18 +40,29 @@ UTF8OutPipeReader::UTF8OutPipeReader(wil::unique_hfile& outPipe) :
...
 
     dwRead += _dwPartialsLen;
     _dwPartialsLen = 0;
 
-    if (dwRead == 0) // quit if no data has been read and no cached data was left over
+    if (!fSuccess) // reading failed (we must check this first, because dwRead will also be 0.)
     {
-        return S_OK;
+        auto lastError = GetLastError();
+        if (GetLastError() == ERROR_BROKEN_PIPE)
```
Comment (german-one) [inline src/types/UTF8OutPipeReader.cpp]: `if (lastError == ERROR_BROKEN_PIPE)` maybe?

Comment (DHowett-MSFT) [reply src/types/UTF8OutPipeReader.cpp]: Whoops, good catch!

**Pull request merged**